### DISCUSSION TITLE
[Feature] Integrate assets and albums shared with a user in this user own timeline

### DISCUSSION THREAD

**Discussion by ghost** [immich-app/immich#3984] [Feature Request]
### The feature

Is it possible to view the assets that other users shared with me as if they were part of my own timeline ?

For exemple, when we do a common event, every participant user upload its assets for this event in the shared album, and every participant would like these assets to be integrated in there own timeline, as if they were the original owners.

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by gsacre** [comment]
I second this!
My daughter is now uploading pictures on my Immich server and she shares the pictures with me, but it'd be nice if they were considered as part of mine as well, so that I can create an album with all our pictures and then share it externally.

I've looked a bit at the tables and came up with a "simple" SQL query that maybe would do the trick (for my test, I simply counted the assets):

```sql
select count(*) 
from assets 
where  "ownerId" = '<my-id>' 
or id in (
    select "assetsId" 
    from albums_assets_assets 
    join albums_shared_users_users on albums_assets_assets."albumsId" = albums_shared_users_users."albumsId" 
           and albums_shared_users_users."usersId" = '<my-id>'
);
```

It's not really optimized and I'm sure someone will find a better way to do this, but I hope it's a starting point.

I've looked at the code to see if I could change it myself, but I'm not good at Javascript. So I believe that this should be adapted in the "getAll" function of the src/infra/repositories/asset.repository.ts file but not 100% sure...

**Reply by jrasm91** [reply]
If you do that though, especially in the web, you now need to adapt the following:
- enable the option to turn it on or off
- handle multi select actions that include assets shared with you (delete, favorite, archive, etc.) Some actions don't make sense anymore. Can you archive an asset from your view that you don't own?
- Hide the favorite button/action, since favoriting is currently only tracked per user (asset owner). 

Displaying them alone isn't too hard, it is more challenging to make everything else still work properly.

**Comment by gsacre** [comment]
> If you do that though, especially in the web, you now need to adapt the following:
> 
>* enable the option to turn it on or off

Agree, this should be an opt-in kind of thing.

>* handle multi select actions that include assets shared with you (delete, favorite, archive, etc.) Some actions don't make sense anymore. Can you archive an asset from your view that you don't own?

I believe they should be considered yours (please see below a proposal for this).

>* Hide the favorite button/action, since favoriting is currently only tracked per user (asset owner).

I could perfectly understand that.

> Displaying them alone isn't too hard, it is more challenging to make everything else still work properly.

I'm sure it wouldn't be an easy task, but could we maybe extract the "ownerID" from the "assets" table?
This way, when a folder would be shared among Immich users of the same instance, a new tuple would be added, giving kind of a copy of the asset to the shared party?
I would see the DB model as this:

```sql
immich=# \d assets
                                   Table "public.assets"
      Column      |           Type           | Collation | Nullable |        Default        
------------------+--------------------------+-----------+----------+-----------------------
 id               | uuid                     |           | not null | uuid_generate_v4()
 deviceAssetId    | character varying        |           | not null | 
 deviceId         | character varying        |           | not null |                  <--- removed ownerID
 type             | character varying        |           | not null | 
...

immich=# \d assets_owners_owners                 <--- new table
                                   Table "public.assets"
      Column      | Type | Collation | Nullable |        Default        
------------------+------+-----------+----------+-----------------------
assetsID          | uuid |           | not null | 
ownersID          | uuid |           | not null | 

```

Then, when displaying the timelines, the query would run on both assets and assets_owners_owners. When sharing, Immich would add  lines in the assets_owners_owners with the same asset IDs as the "origin" pictures but the "target" owner. Then the pictures would have their own lives in the "target" owner pictures (favoriting, archiving, deleting but keeping the file as long as there is one entry in assets_owners_owners...).
The only question would be "what if the original picture owner decides to stop sharing"? Then I believe we could retrieve the list of asset IDs from the albums_assets_assets table and delete the entries in assets_owners_owners.

Obviously I cannot even imagine the danger of taking "ownerID" out of "assets" as I cannot properly read the code and evaluate all the implications (I wish I could submit a merge request for this if it was written in a coding language I knew).
But if this makes sense to you, maybe would solve my itch and potentially others :-)

**Reply by jrasm91** [reply]
Yes, something like this would be ideal. Currently `isFavorite` lives on the asset table though, so it is only tracked _per asset_, not _per user per asset_. In your example you would need to move that column to the intersection table between assets and owners. Same with any other _per user_ properties. As far as I understand this is fairly complicated, but obviously not impossible. I just wanted to point out that these are the considerations that make it more complicated to work on. In the web at least, there is a lot of logic built around whether `asset.ownerId` is the current owner or not.

FWIW there is a some overlap with "groups", which is another feature request which also abstracts asset ownership a bit.

**Comment by NickyM** [comment]
This! I can't even explain how staggered I was, that when I finally got my Google takeout takeout imported and shared way-to-many albums with my SO and then realised that this wasnt an option.

In Evil Companys Photos, it is possible to toggle this for each album. Eg, i want all pics of "Christmas 2022" included in my timeline, search results, facial recognition etc, but I do no want "Weekend with SOs extended family 2021" displayed.

Otherwise this software is so impressive, that my arms is still cant come down ;)

**Comment by mirisbowring** [comment]
I love this! Also an option to enable the inclusion of Partner Sharing into the own Timeline would be great!

Since we became parents this year, we are taking many photos and we don't want o have a shared "baby" album and manually add those images into it.

**Comment by rexkani** [comment]
I second this. Currently sharing albums with family members and they are only viewers, having such feature can make their app experience more organic

**Comment by ptfuzi** [comment]
This would be nice, something that it’s missing, in my perspective

**Comment by bo0tzz** [comment]
#1779
